### ISSUE TITLE
Figure out of COMPILE, should be flagged NN

### ISSUE THREAD

**Issue by scotws** [scotws/TaliForth2#123]
@SamCoVT wrote in a different issue: "Shouldn't COMPILE_COMMA be flagged NN? It has an absolute jump in it, so if it were native compiled, the new version would jump back into the old version in the _compile_as_jump case. It won't hurt anything if that happens, but I was under the impression that words with JMP (as opposed to BRA) instructions in them would be NN. If that's true, you may want to just flag COMPILE_COMMA as NN in the headers."

I might have to think about this one, because we should have a general rule about these things, and I'm not sure if I had one way back when with Tali 1 or it just looks like I knew what I was doing. Saving this as a separate issue until I can come back to this.

**Comment by SamCoVT** [comment]
If a word uses a JMP instruction in it, then it's rather silly to native compile it as that will create a second copy of the code, but the JMP will still have the original address, so it will still be using part of the original word (if the JMP is taken).  That seems like a waste of memory.

Trying to update the JMP instruction target doesn't make sense because some words will have data in them, and the data might look like a JMP instruction when it's not, similar to how the disasm word gets confused when there is data in a word.  I think trying to update JMPs is a bad idea.

Words that use flow control are going to have JMP instructions in them.

I see that new words are marked NN by default.  I think this is a good policy, and perhaps we should add to the documentation that `always-native` should only be called on user-defined words that have no flow control in them.

My proposal is the following:
1.) Add the NN flag to `COMPILE,`.  I've done a search (by looking for "jmp _") and did not find any other words in this situation other than `ACCEPT`, which already has the NN flag.
2.) Update the documentation to say that `always-native` is discouraged on user words that use flow control, such as IF, CASE, or any of the loops, along with the reason why.  We don't need to forbid the user from using `always-native` on these words because the side effect (starts in the natively compiled word but may jump to the original word) is harmless other than wasting memory and it is technically still faster by eliminating the JSR to the original word.

**Comment by SamCoVT** [comment]
I originally though this would be harmless, but after noodling on it some more, I don't think it's harmless anymore.  The issue is that, when native compiled, the code is copied over _and the RTS is stripped_.  If a new word (let's call is AWORD) is defined and a word with a jump is native-compiled into it, then when AWORD is run, it will jump back into the original word, and AWORD will return when it hits the RTS of the original word.  Unless the native-compiled word was the very last one, this RTS will prevent the rest of the code in AWORD from running.

I've make a test to prove this is the case.  
```
: testnative 3 0 do i . loop ; always-native  ok
: aword ." Before... " testnative ." After!" ;  ok
aword Before... 0 1 2  ok
```
It appears my extended noodling came up with the correct conclusion, which is a bit unfortunate as I liked this issue better when it was harmless.

I'll fix up `COMPILE,` and put an explanation of when you shouldn't use `always-native` and why.  I believe `COMPILE,` was the only word left that had a JMP in it but was marked always-native.  I'll double check that as well.

You get bonus points for making all new words never-native by default (I always like sensible defaults,) which made this much less of an issue than it could have been.

**Comment by SamCoVT** [comment]
I looked at every "jmp" in native_words.asm.  Ignoring the jumps to error or underflow, I found the jmp instruction in the following words that were not currently marked NN:
`ABORT" COMPILE, PARSE-NAME SEARCH`
The words that currently have a jmp but are already marked NN are:
`ACCEPT MOVE`
and the following words have a jmp but are doing trickery that I believe to be native-compiling-safe:
`EMIT KEY EXECUTE`
You may want to look at that last list just to make sure those words really are save to native-compile.

ABORT" is a little weird, as it's a compile-only immediate word, but I figured that adding the never-native flag couldn't hurt, and for all I know someone might POSTPONE it into another word.

I've added the NN flag to all of the words in the first list, and it all passes the test suite.  I'll work up some documentation on always-native and then issue a pull request for it all.

**Comment by scotws** [comment]
Faced with the idea of adding _another_ word for native compile, I'm going to take a step back and ask if we really think this whole mechanism is worth it. Yes, it reduces `jsr xt_drop` to `inx inx` etc, which makes a lot of words faster and shorter. On the other hand, we're using a whole bunch of space for all these words and it makes everything a lot more complicated, error-prone, and harder to understand. 

Or can we think of a better mechanism for this? One of the alternatives I had originally considered is a separate "optimizer" program that takes the xt of a word and then walks through it, replacing certain subroutine jumps with inline versions:
```
' myword optimize
```
The program would add a new, optimized version of the word to the Dictionary. 

At the time, this sounded like a scary amount of analysis and copying stuff around, but now I'm not so sure. It would get rid of a bunch of special words, make defining new words dirt simple, and we could have a ball writing a tool to optimize 65c02 code.

**Comment by scotws** [comment]
Yeah, that's they scary part I was talking about - I actually read up some background on code optimization in Cooper and Torczon's _Engineering a Compiler_ but started to feel ill when I got to "data-flow analysis". By the time I got to "scalar optimizations", I had to go pet the cat for an hour till I felt better again. 

Okay, your suggestions make sense, and you are of course right that the code has had so much testing by now that it would be foolish to kick it out. I'd be grateful if you could make the suggested changes. But, for the record, my next Forth is going to do this differently :-).
